app/src/main.c

```diff
@@ -0,0 +1,56 @@
+/**
+ * @author Nordic Semiconductor ASA
+ * @date 2021
+ * @copyright Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <drivers/sensor.h>
+
+#include "app_version.h"
+#include "foo.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(main, CONFIG_APP_LOG_LEVEL);
+
+int main(void)
+{
+	int ret;
+	const struct device *sensor;
+
+	printk("Zephyr Example Application %s\n", APP_VERSION_STR);
+
+	ret = app_foo_bar();
+	if (ret < 0) {
+		LOG_ERR("app_foo_bar failed (%d)", ret);
+		return ret;
+	}
+
+	sensor = DEVICE_DT_GET(DT_NODELABEL(examplesensor0));
+	if (!device_is_ready(sensor)) {
+		LOG_ERR("Sensor not ready");
+		return -ENODEV;
+	}
+
+	while (1) {
+		struct sensor_value val;
+
+		ret = sensor_sample_fetch(sensor);
+		if (ret < 0) {
+			LOG_ERR("Could not fetch sample (%d)", ret);
+			return ret;
+		}
+
+		ret = sensor_channel_get(sensor, SENSOR_CHAN_PROX, &val);
+		if (ret < 0) {
+			LOG_ERR("Could not get sample (%d)", ret);
+			return ret;
```

Comment:
IIRC, the Zephyr initialization process does not include either a structured teardown or registration of assert handlers.  This means returning from `main()` would not terminate the OS. So ... why return from main at all?

Reply:
> So ... why return from main at all?

Maybe because if this fails, there's nothing else for main to do? What would you suggest; an infinite loop?

Reply:
The first problem is that the function signature available in Zephyr is `void main(void);`, so returning any value is a coding error.

The second problem is not distinguishing between catastrophic failures, degraded functionality, vs temporarily unavailable functionality.  Different recoveries are needed for each.  Catastrophic failures that indicate the executable environment is untrustworthy need a device reset to recover (akin to an assert()).  Detecting degraded functionality need to update some query-able status.  Temporarily unavailable functionality should not terminate the thread (because we cannot reinitialize/restart it).

Reply:
ISO C states that the main function returns `int`, so I'd not use `void main`. When using `int main` we need to return something, even if it's later not used by Zephyr.